data/store/mongo/mongo.go

```diff
@@ -913,3 +925,273 @@ func validateAndTranslateSelectors(selectors *data.Selectors) (bson.M, error) {
 
 	return selector, nil
 }
+
+// assumes all except freestyle is 5 minutes
+func getDuration(row *continuous.Continuous) int64 {
+    if strings.Contains(*row.DeviceID, "AbbottFreeStyleLibre") {
+        return 15
+    }
+    return 5
+}
+
+func (d *DataRepository) CalculateSummary(ctx context.Context, summary *data.Summary) (*data.Summary, error) {
+	if ctx == nil {
+		return nil, errors.New("context is missing")
+	}
+
+	if summary == nil {
+		return nil, errors.New("original summary is missing")
+	}
+
+	if summary.UserID == "" {
+        return nil, errors.New("summary missing UserID")
+    }
+
+	id := summary.UserID
+
+	var dataSet []*continuous.Continuous
+	selector := bson.M{
+		"_active": true,
+		"_userId": id,
+        "type": "cbg",
+	}
+
+    findOptions := options.Find()
+    findOptions.SetSort(bson.D{{"time", -1}})
+    findOptions.SetLimit(1)
+
+	cursor, err := d.Find(ctx, selector, findOptions)
+
+    if err != nil {
+		return nil, errors.Wrap(err, "unable to get last upload date")
+	}
+
+	if err = cursor.All(ctx, &dataSet); err != nil {
+		return nil, errors.Wrap(err, "unable to decode last upload date")
+	}
+
+    lastUpload, err := time.Parse(time.RFC3339Nano, *dataSet[0].CreatedTime)
+    endTime, err := time.Parse(time.RFC3339Nano, *dataSet[0].Time)
+
+    // remove 2 weeks for start time
+    startTime := endTime.AddDate(0, 0, -14)
+
+    // if last LastData date is more recent than 2 weeks, try rolling calc
+    var weight float64 = 1.0
+    if startTime.Before(summary.LastData) {
+        // get ratio between start time and actual start time for weights
+        existingTime := summary.LastData.Sub(startTime)
+        newTime := endTime.Sub(summary.LastData)
+        weight = newTime.Seconds()/existingTime.Seconds()
+
+        startTime = summary.LastData
+    }
+
+    var dataSets []*continuous.Continuous
+    selector = bson.M{
+		"_active": true,
+		"_userId": id,
+        "type": "cbg",
+        // we currently pull all so we can do in-range and avg in 1 pass, maybe bad idea
+        //"value": bson.M{"$gt": 3.9, "$lt": 10},
+        "time": bson.M{"$gt": startTime.Format(time.RFC3339Nano),
+                       "$lte": endTime.Format(time.RFC3339Nano)},
+	}
+
+    cursor, err = d.Find(ctx, selector)
+
+    if err != nil {
+		return nil, errors.Wrap(err, "unable to get in-range values")
+	}
+
+    if err = cursor.All(ctx, &dataSets); err != nil {
+		return nil, errors.Wrap(err, "unable to decode data sets for user by id")
+	}
+
+    totalMinutes := float64(math.Round(endTime.Sub(startTime).Minutes()))
+
+    var inRangeMinutes int64 = 0
+    var belowRangeMinutes int64 = 0
+    var aboveRangeMinutes int64 = 0
+    var totalGlucose float64 = 0
+
+    for _, r := range dataSets {
+        if *r.Value < 3.9 {
```

Comment:
Consider making these named constants even if they are universally accepted. Naked numbers always leave room for someone not familiar with diabetes to go "huh, whazzat?"